File: algorithms/11_golang/numerical/monte_carlo_integration.go
package main

import (
	"fmt"
	"math"
)

func monteCarloIntegration(f func(float64) float64, a, b float64, samples int) float64 {
	state := int64(42)
	modulus := int64(1) << 31
	total := 0.0
	for i := 0; i < samples; i++ {
		state = (1103515245*state + 12345) % modulus
		total += f(a + (b-a)*float64(state)/float64(modulus))
	}
	return (b - a) * total / float64(samples)
}

func main() {
	estimate := monteCarloIntegration(func(x float64) float64 { return x * x }, 0, 1, 100000)
	if math.Abs(estimate-1.0/3.0) > 0.01 {
		panic("monte carlo failed")
	}
	fmt.Println("monte carlo integration ok")
}

File: algorithms/06_javascript/numerical/monte_carlo_integration.js
function monteCarloIntegration(f, a, b, samples = 100000, seed = 42) {
  let state = seed;
  let total = 0;
  const modulus = 2 ** 31;
  for (let i = 0; i < samples; i += 1) {
    state = (1103515245 * state + 12345) % modulus;
    total += f(a + (b - a) * (state / modulus));
  }
  return ((b - a) * total) / samples;
}

module.exports = { monteCarloIntegration };

if (require.main === module) {
  const estimate = monteCarloIntegration((x) => x * x, 0, 1);
  if (Math.abs(estimate - 1 / 3) > 0.01) throw new Error("monte carlo integration failed");
  console.log("[JavaScript Monte Carlo Integration] verified");
}
